### DISCUSSION TITLE
Transpilation for Real Hardware

### DISCUSSION THREAD

**Discussion by plutoniumm** [Qiskit-Extensions/circuit-knitting-toolbox#457] [Q&A] [answered]
I am currently playing with wire cutting and want to be able to run it on real hardware
I am also NOT using CutQC to automatically calculate the cuts for me which I know will I think resolve most of the below issues

The objective is to be able to transpile the circuit for a given backend
So say if I
- create a `RealAmplitudes` (6 qubits)
- Cut it at wire 2 (Manual insertion of `CutWire` & then call `cut_wires` from v0.5.0)
- Pass EACH of them to `transpile` with `backend="ibmq_kolkata"`

It seems to error out in the following ways
- If I call `.decompose()` on any ONE of the cuts: `ValueError: Missing 'basis_id': unable to realize SingleQubitQPDGate` OR
- If I just pass to Transpiler without decomposition: `'qpd_1q', 'rz', 'ryy', 'u3', 'rxx'] to the backend's (or manually specified) target basis: ['delay', 'snapshot', 'sx', 'rz', 'cx', 'measure', 'id', 'x', 'barrier', 'reset’]`
	Caveat: basis_gates=backend.config…basis gates (from same backend above i.e ibmq_kolkata)

The overview of what im trying is
Take a Circuit → Cut it → Transpile both for given backend → Apply mapomatic → Run

Easiest way to replicate would be
- [Use this circuit](https://qiskit-extensions.github.io/circuit-knitting-toolbox/circuit_cutting/how-tos/how_to_specify_cut_wires.html)
- Make the cuts, i.e we should now have `subcircuit[0]` and `subcircuit[1]`
- `backend=provider.get_backend('ibmq_kolkata')`
- `transpile(subcircuit[0], backend=backend, optimization_level=3)`
- (I never got here but) then run Mapomatic and run Sampler on each of them

**Comment by garrison** [comment, marked as answer]
I believe the step you are missing is to run `generate_cutting_experiments` to generate the explicit cutting experiments, as demonstrated in [the first tutorial](https://qiskit-extensions.github.io/circuit-knitting-toolbox/circuit_cutting/tutorials/01_gate_cutting_to_reduce_circuit_width.html).

That said, once you get past the current error, I expect you may also experience the an error about resets, given in #452.  I've provided a workaround in that issue and am working on a PR that fixes it in CKT.

If you follow the above suggestion, you will need to run mapomatic for each transpiled experiment.  If you'd like to run mapomatic once before generating the subexperiments, this is likely possible too, but it will be more involved (I can provide some guidance; in short, the test [here](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/blob/190552d955ac5017d6a60967535cc1159ba4ec6f/test/cutting/test_cutting_workflows.py#L33) demonstrates how to pass a `SingleQubitQPDGate` through the transpiler).

**Reply by plutoniumm** [reply]
Hey thanks a lot It seems this works, It seems like for mapomatic i have one of two options
1. Check for `subexperiments[0][0]` what the best backend is (testing against all of `provider.backends()` ASSUMING all experiments will run best on that backend
2. Check for ALL 100s of experiments and see what is best

Right now I'm gonna sample a 10-10 circuits a few times and poll those to see what is best backend

This however has the implication that we CANNOT find the best backend BEFORE `generate_cutting_experiments`. While it makes sense why (since circuits may change), it may be worthwhile to think of a way to do so before to reduce computational load and just do it once and fix the backend for that subcircuit
